File: components/file_scanner/src/kohighlights/file_scanner/scanner.py
"""Filesystem scanner for KOReader ``.sdr`` metadata folders.

Usage::

    from kohighlights.file_scanner import scan_for_books

    for book in scan_for_books("/mnt/kobo", on_progress=print):
        print(book.title, book.highlight_count)

The scanner walks the given root directory looking for ``*.sdr`` folders that
contain a ``metadata.*.lua`` file, then decodes the Lua file into a
:class:`~kohighlights.models.Book`.
"""
from __future__ import annotations

import hashlib
import os
from datetime import datetime
from os.path import getmtime, isfile, join, splitext, basename
from pathlib import Path
from typing import Callable, Generator, Iterable

from kohighlights.lua_codec import decode_file
from kohighlights.models import Book, Highlight


# ── Helpers ──────────────────────────────────────────────────────────────────


def _md5_short(path: str) -> str:
    """Return the first 8 chars of the MD5 hash of the file path string."""
    return hashlib.md5(path.encode("utf-8")).hexdigest()[:8]


def get_book_path(meta_path: str, data: dict) -> str:
    """Derive the actual book file path from the ``.sdr`` metadata path.

    KOReader stores metadata in ``<bookfile>.sdr/metadata.<ext>.lua``.
    We reverse-engineer the original book path from that.
    """
    # meta_path: /path/to/Book Title.epub.sdr/metadata.epub.lua
    sdr_dir = os.path.dirname(meta_path)               # …/Book Title.epub.sdr
    parent_dir = os.path.dirname(sdr_dir)              # …/
    sdr_name = os.path.basename(sdr_dir)               # Book Title.epub.sdr
    if sdr_name.endswith(".sdr"):
        book_name = sdr_name[:-4]                      # Book Title.epub
        return join(parent_dir, book_name)

    # Fallback: try doc_path from metadata
    doc_path = data.get("doc_path", "")
    return doc_path if doc_path else ""


def _extract_highlights(raw: dict) -> list[Highlight]:
    """Convert the raw ``highlight`` Lua dict into a list of :class:`Highlight`."""
    raw_highs = raw.get("highlight", {})
    if not raw_highs:
        return []

    highlights: list[Highlight] = []

    # KOReader stores highlights as nested dicts: {page: {idx: {text, …}}}
    if isinstance(raw_highs, dict):
        for page_key, page_data in raw_highs.items():
            if not isinstance(page_data, dict):
                continue
            for _idx, h in page_data.items():
                if not isinstance(h, dict):
                    continue
                text = str(h.get("text", "")).strip()
                if not text:
                    continue
                # Find a comment if there is an associated note
                comment = ""
                bookmark_data = raw.get("bookmarks") or {}
                # Some versions store the comment alongside the highlight
                comment = str(h.get("note", "") or h.get("comment", "")).strip()

                highlights.append(Highlight(
                    text=text,
                    page=int(page_key) if str(page_key).isdigit() else 0,
                    date=str(h.get("datetime", "") or h.get("date", "")).strip(),
                    chapter=str(h.get("chapter", "")).strip(),
                    comment=comment,
                    page_id=str(h.get("pageno", "") or h.get("pos0", "")).strip(),
                ))
    elif isinstance(raw_highs, list):
        # Newer KOReader format: flat list
        for h in raw_highs:
            if not isinstance(h, dict):
                continue
            text = str(h.get("text", "")).strip()
            if not text:
                continue
            highlights.append(Highlight(
                text=text,
                page=int(h.get("page", 0) or 0),
                date=str(h.get("datetime", "") or h.get("date", "")).strip(),
                chapter=str(h.get("chapter", "")).strip(),
                comment=str(h.get("note", "") or h.get("comment", "")).strip(),
                page_id=str(h.get("pageno", "") or h.get("pos0", "")).strip(),
            ))

    return highlights


def _raw_to_book(meta_path: str, raw: dict) -> Book:
    """Convert a decoded Lua dict into a :class:`Book`."""
    stats = raw.get("stats", {}) or {}
    summary = raw.get("summary", {}) or {}

    title = (stats.get("title") or raw.get("title") or "").strip()
    authors = (stats.get("authors") or raw.get("authors") or "").strip()

    try:
        pct = float(raw.get("percent_finished", 0) or 0)
    except (ValueError, TypeError):
        pct = 0.0

    rating_num = summary.get("rating")
    rating = (str(rating_num) + "★") if rating_num else ""

    modified = str(datetime.fromtimestamp(getmtime(meta_path))).split(".")[0]

    return Book(
        path=meta_path,
        title=title or "NO TITLE FOUND",
        authors=authors or "NO AUTHOR FOUND",
        series=str(raw.get("series", "") or "").strip(),
        language=str(raw.get("language", "") or "").strip(),
        pages=int(stats.get("pages", 0) or raw.get("pages", 0) or 0),
        keywords=str(raw.get("keywords", "") or "").strip(),
        highlights=_extract_highlights(raw),
        rating=rating,
        status=str(summary.get("status", "") or "").strip(),
        percent_finished=pct,
        modified_date=modified,
        cre_dom_version=str(raw.get("cre_dom_version", "") or "").strip(),
        md5=str(raw.get("partial_md5_checksum", _md5_short(meta_path))),
        book_path=get_book_path(meta_path, raw),
        raw_data=raw,
    )


def _find_lua_files(root: str) -> Generator[str, None, None]:
    """Yield absolute paths to ``metadata.*.lua`` files under *root*."""
    for dirpath, dirnames, filenames in os.walk(root, followlinks=True):
        # KOReader places metadata inside *.sdr directories
        if dirpath.endswith(".sdr"):
            for fn in filenames:
                if fn.startswith("metadata.") and fn.endswith(".lua"):
                    yield join(dirpath, fn)
        # Prune hidden directories to speed up scans on large trees
        dirnames[:] = [d for d in dirnames if not d.startswith(".")]


# ── Public API ───────────────────────────────────────────────────────────────


def scan_for_books(
    roots: str | Iterable[str],
    on_progress: Callable[[str], None] | None = None,
    skip_paths: set[str] | None = None,
) -> Generator[Book, None, None]:
    """Walk *roots* for KOReader metadata and yield :class:`Book` objects.

    Args:
        roots:       A single path or iterable of paths to scan.
        on_progress: Optional callback called with each ``meta_path`` found,
                     before it is decoded.  Useful for progress reporting.
        skip_paths:  Set of ``meta_path`` values to skip (already loaded).

    Yields:
        :class:`~kohighlights.models.Book` instances.
    """
    if isinstance(roots, str):
        roots = [roots]

    skip = skip_paths or set()

    for root in roots:
        root = str(root)
        if not os.path.exists(root):
            continue
        # Accept a direct path to a .lua file or a directory to scan
        if isfile(root) and root.endswith(".lua"):
            lua_files: Iterable[str] = [root]
        else:
            lua_files = _find_lua_files(root)

        for meta_path in lua_files:
            if meta_path in skip:
                continue
            if on_progress:
                on_progress(meta_path)
            raw = decode_file(meta_path)
            if not raw:
                continue
            try:
                yield _raw_to_book(meta_path, raw)
            except Exception:
                continue  # skip malformed entries silently


def scan_from_db_records(
    records: Iterable[tuple[str, str, str, dict]],
) -> Generator[Book, None, None]:
    """Convert archived DB records into :class:`Book` objects.

    Args:
        records: Iterable of (md5, path, date, data_dict) tuples as returned
                 by :meth:`~kohighlights.book_store.BookStore.all_records`.

    Yields:
        :class:`~kohighlights.models.Book` instances.
    """
    for md5, path, date, data in records:
        try:
            book = _raw_to_book(path, data)
            book.md5 = md5
            book.modified_date = date
            yield book
        except Exception:
            continue

File: tauri/KoHighlights/src-tauri/src/scanner.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;
use md5;

use crate::error::Result;
use crate::lua::decode_file;
use crate::models::{Book, Highlight};

/// Scan directory for KOReader .lua metadata files and yield Books
pub fn scan_for_books(root: &Path) -> Result<Vec<Book>> {
    let mut books = Vec::new();

    for entry in WalkDir::new(root)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| !e.path().components().any(|c| {
            if let std::path::Component::Normal(n) = c {
                n.to_string_lossy().starts_with('.')
            } else {
                false
            }
        }))
    {
        let path = entry.path();

        // Look for *.sdr/metadata.*.lua files
        if path.file_name()
            .and_then(|n| n.to_str())
            .map(|n| n.starts_with("metadata.") && n.ends_with(".lua"))
            .unwrap_or(false)
        {
            // Check parent is .sdr directory
            if let Some(parent) = path.parent() {
                if parent
                    .file_name()
                    .and_then(|n| n.to_str())
                    .map(|n| n.ends_with(".sdr"))
                    .unwrap_or(false)
                {
                    match decode_file(path) {
                        Ok((raw_data, original_header)) => {
                            if let Ok(book) = raw_to_book(path, raw_data, original_header) {
                                books.push(book);
                            }
                        }
                        Err(_) => continue,
                    }
                }
            }
        }
    }

    Ok(books)
}

/// Convert raw Lua dict to Book with highlights extracted
fn raw_to_book(path: &Path, raw_data: serde_json::Value, original_header: String) -> Result<Book> {
    let obj = raw_data.as_object().ok_or_else(|| {
        crate::error::AppError::Logic("raw_data must be object".to_string())
    })?;

    // Extract metadata fields from raw_data
    let title = obj
        .get("title")
        .or_else(|| obj.get("stats").and_then(|s| s.get("title")))
        .and_then(|v| v.as_str())
        .unwrap_or("NO TITLE FOUND")
        .to_string();

    let authors = obj
        .get("authors")
        .or_else(|| obj.get("stats").and_then(|s| s.get("authors")))
        .and_then(|v| v.as_str())
        .unwrap_or("NO AUTHOR FOUND")
        .to_string();

    let series = obj
        .get("series")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let language = obj
        .get("language")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let pages = obj
        .get("pages")
        .or_else(|| obj.get("stats").and_then(|s| s.get("pages")))
        .and_then(|v| v.as_i64())
        .unwrap_or(0) as i32;

    let keywords = obj
        .get("keywords")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let rating = obj
        .get("summary")
        .and_then(|s| s.get("rating"))
        .and_then(|v| v.as_str())
        .map(|s| format!("{}★", s))
        .unwrap_or_default();

    let status = obj
        .get("summary")
        .and_then(|s| s.get("status"))
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let percent_finished = obj
        .get("percent_finished")
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);

    let cre_dom_version = obj
        .get("cre_dom_version")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    // Compute MD5 from path (fallback if partial_md5_checksum not in raw_data)
    let md5 = obj
        .get("partial_md5_checksum")
        .and_then(|v| v.as_str())
        .unwrap_or_else(|| {
            let digest = md5::compute(path.to_string_lossy().as_bytes());
            let hex = format!("{:x}", digest);
            &hex[..8.min(hex.len())]
        })
        .to_string();

    let modified_date = fs::metadata(path)
        .ok()
        .and_then(|m| m.modified().ok())
        .and_then(|t| {
            let duration = t
                .duration_since(std::time::UNIX_EPOCH)
                .ok()?;
            let secs = duration.as_secs() as i64;
            chrono::DateTime::<chrono::Utc>::from_timestamp(secs, 0).ok()
        })
        .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string())
        .unwrap_or_default();

    let book_path = path
        .parent()
        .and_then(|p| p.file_name())
        .and_then(|n| n.to_str())
        .map(|s| s.strip_suffix(".sdr").unwrap_or(s).to_string())
        .unwrap_or_default();

    // Extract highlights from raw_data
    let highlights = extract_highlights(obj);

    Ok(Book {
        path: path.to_string_lossy().to_string(),
        title,
        authors,
        series,
        language,
        pages,
        keywords,
        highlights,
        rating,
        status,
        percent_finished,
        modified_date,
        cre_dom_version,
        md5,
        book_path,
        raw_data,
        original_header,
    })
}

fn extract_highlights(obj: &serde_json::Map<String, serde_json::Value>) -> Vec<Highlight> {
    let mut highlights = Vec::new();

    if let Some(highlight_data) = obj.get("highlight") {
        // Dict format: {page: {idx: {text, datetime, chapter, note, pageno}}}
        if let Some(page_dict) = highlight_data.as_object() {
            for (page_key, page_val) in page_dict {
                let page = page_key.parse::<i32>().unwrap_or(0);

                if let Some(idx_dict) = page_val.as_object() {
                    for (_idx, h_val) in idx_dict {
                        if let Some(h_obj) = h_val.as_object() {
                            let text = h_obj
                                .get("text")
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();

                            let date = h_obj
                                .get("datetime")
                                .or_else(|| h_obj.get("date"))
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();

                            let chapter = h_obj
                                .get("chapter")
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();

                            let comment = h_obj
                                .get("note")
                                .or_else(|| h_obj.get("comment"))
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();

                            let page_id = h_obj
                                .get("pageno")
                                .or_else(|| h_obj.get("pos0"))
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();

                            highlights.push(Highlight {
                                text,
                                page,
                                date,
                                chapter,
                                comment,
                                page_id,
                            });
                        }
                    }
                }
            }
        }
    }

    // Also check for array format (newer KOReader)
    if let Some(highlight_array) = obj.get("highlights") {
        if let Some(arr) = highlight_array.as_array() {
            for h_val in arr {
                if let Some(h_obj) = h_val.as_object() {
                    let text = h_obj
                        .get("text")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let page = h_obj.get("page").and_then(|v| v.as_i64()).unwrap_or(0) as i32;

                    let date = h_obj
                        .get("datetime")
                        .or_else(|| h_obj.get("date"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let chapter = h_obj
                        .get("chapter")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let comment = h_obj
                        .get("note")
                        .or_else(|| h_obj.get("comment"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let page_id = h_obj
                        .get("pageno")
                        .or_else(|| h_obj.get("pos0"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    highlights.push(Highlight {
                        text,
                        page,
                        date,
                        chapter,
                        comment,
                        page_id,
                    });
                }
            }
        }
    }

    highlights
}
